### PR TITLE
Ngram classify

### PR DESCRIPTION
**Pull request by AntonGorokhov**
### Changelog category:
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Ngram Classification function added, so that user can SELECT smth with it

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by AntonGorokhov** [comment]
Hello, this is my pull request. Not everything is done yet.
I am about to finish it in couple of days

File: src/Functions/FunctionsNgramClassify.h [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        auto col_res = ColumnString::create();
+
+        // const ColumnString * col_vector = checkAndGetColumn<ColumnString>(&*arguments[1].column);
+
+        // const auto &chars = col_vector->getChars();
+        // const auto &offsets = col_vector->getOffsets();
+
+        // const String &slice = slice_name_const->getValue<String>();
+
+        // Impl::vector(chars, offsets, slice, col_res->getChars(), col_res->getOffsets());
```
Comment (qoega) [inline src/Functions/FunctionsNgramClassify.h:104]: Drop it?

**Review by qoega — COMMENTED** [review]
Is it still work in progress? Do you plan to add perf and stateless tests?

File: .zshrc
[the remark this answers is not in the document]
Comment (AntonGorokhov) [reply .zshrc:1]: Yes, added accidentally

**Comment by AntonGorokhov** [comment]
> 

Yes, work is still in progress

**Comment by robot-ch-test-poll2** [comment]
This is an automated comment for commit 8ba2fd28ee58a195d304c28d3f4eb577f777d859 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/50247/8ba2fd28ee58a195d304c28d3f4eb577f777d859/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>🟡 pending</td></tr>
</table>

File: src/Functions/FunctionsNgramClassify.cpp [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+    3. Сливаем модельки -> в каждую хешмапу добавляем все уникальные ngram-ы другой
+    4. Нормируем мапы 
+    5. Считаем формулу Байеса (конечно же с логарифмами)
+    6. Возвращаем вероятность принадлежности строки к модели
+
+И дальше в зависимости от того, что нужно дальше 
+    - Либо просто нормируем вероятности и возвращаем вероятность пренадлежности нашему классу
+    - Либо сортируем всех (отнормированных) и возвращаем вектором пар
+    - Либо просто возвращаем наиболее вероятную модельку
+*/
```
[the remark this answers is not in the document]
Comment (AntonGorokhov) [reply src/Functions/FunctionsNgramClassify.cpp:57]: Done

File: opt/NgramModels/Books/sherlock.txt
[the remark this answers is not in the document]
Comment (AntonGorokhov) [reply opt/NgramModels/Books/sherlock.txt:1]: These files are going to be used as Ngram Models inside ClickHouse. What should I do to register them?

**Comment by AntonGorokhov** [comment]
Please explain me how to declare /opt directory with models that I added in a proper way

File: opt/NgramModels/songs-ru/Oxxxymiron [hunk 3]
```diff
@@ -0,0 +1,72 @@
+Дон ли, Волга ли течёт котомку на плечо
```
Comment (alexey-milovidov) [inline opt/NgramModels/songs-ru/Oxxxymiron:1]: That's probably copyrighted, and I don't think it's fair use.

**Comment by alexey-milovidov** [comment]
This is not as discussed. The model file should contain (class_id, ngram, count) triples.

**Pull request closed without merging**